Show abstract states in program source code

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 7258f1576f643f4a69a769bc5be5d4b0603f47f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88660117

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,156 @@
+#include <fstream>
+
+/// get a path to the file in given source location
+static std::string file(const source_locationt &l)
```
I think this utility function should be moved to `util/source_location.{cpp,h}` as 1) it may be useful in other places and 2) it builds on information that is specific to `source_locationt`, and not to what's happening in this current file.

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,156 @@
+#include <fstream>
+
+/// get a path to the file in given source location
+static std::string file(const source_locationt &l)
```
I think this utility function should be moved to `util/source_location.{cpp,h}` as 1) it may be useful in other places and 2) it builds on information that is specific to `source_locationt`, and not to what's happening in this current file.

kroening replied:
done

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,156 @@
+#endif
+
+  if(!in)
+    return; // we fail silently
```
Why is this ok to do? Shouldn't this be a fatal error?

kroening replied:
I don't want this to be a blocker to viewing -- warning instead?

tautschnig replied:
But then you can't actually view anything, can you? I believe there should be some message to the user that something has gone wrong.

kroening replied:
The idea is that only some of the files might not be there, and you don't want to block viewing the others. Will turn into a warning().

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,156 @@
+  if(!in)
+    return; // we fail silently
+
+  std::map<unsigned, ai_baset::locationt> line_map;
```
I believe this should be `std::size_t` instead of `unsigned` - to me, the only reason to ever use `unsigned` are cases where memory consumption matters and we are damn sure it will be large enough.

kroening replied:
ok

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,156 @@
+
+  unsigned line_no = 0;
+  std::string line;
+  while(std::getline(in, line))
```
Wouldn't this make for a perfectly good `for` loop instead of declaring `line_no` separately?

kroening replied:
ok

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 77e7459e3894085a311ac557c01310c18b6492a8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88822020
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 70269f5f9c3a1fa2f95b47a1f8e9cdc67901481c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88836441

tautschnig commented on src/util/source_location.h:
```diff
@@ -185,6 +185,9 @@ class source_locationt:public irept
     return static_cast<const source_locationt &>(get_nil_irep());
   }
 
+  /// get a path to the file, including working directory
```
This is is different comment string from what's placed with the implementation, I'm not sure what Doxygen is going to do about this. The guidelines say that we only place documentation with the implementation.

kroening replied:
done. Not sure I agree with the guidelines. People read the class.

tautschnig replied:
Ack - how about submitting a PR that proposes a change `CODING_GUIDELINES.md`? I'd say that's the way to instigate change.

tautschnig commented on src/util/source_location.cpp:
```diff
@@ -71,6 +71,20 @@ void source_locationt::merge(const source_locationt &from)
   }
 }
 
+/// get a path to the file in given source location
```
I don't think this makes very clear what this method does. How about:
```
Get a path to the file, including working directory.
\return Full path unless the file name is empty or refers to a built-in, in which case the empty string is returned.
```

kroening replied:
done

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,145 @@
+      const auto line = i_it->source_location.get_line();
+
+      if(line.empty())
+        continue; // ignore
```
There is a lot of overlap between this code and `get_source_files`. Could this be refactored?

kroening replied:
done

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 88b5773949e6f6c6bb5823cf28f379e5ae4ca375).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89081638

tautschnig commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,142 @@
+
+#include <fstream>
+
+static bool show_location(const ai_baset &ai, ai_baset::locationt loc)
```
All callers re-do `.full_path()`, which is non-trivial. How about returning an `optionalt<std::string>` so that this can be avoided? As an added value, that would also better fit the procedure name I'd say.

kroening replied:
Done!

tautschnig reviewed (commented):
One suggestion below, but my main remaining concern is the lack of tests. This is a non-trivial feature that's bound to break eventually if we don't have a single test in CI.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d153f4646a6251b1f5aa091e1d8010757d2c0781).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89314029

kroening commented:
Now with test!

kroening committed:
added show_on_source for abstract domains

The abstract states are shown in a different color on the source code text
of the program, as opposed to the goto-program dump.  Indentation of the
output of the abstract is made to match the indentation of the source code.

tautschnig commented on regression/goto-analyzer/show-on-source/main.c:
```diff
@@ -0,0 +1,12 @@
+  int j;
+  _Bool nondet;
+
+  if(j>=1 && j<=10)
```
Clang-format is complaining about this line.

tautschnig reviewed (approved):
Looks ok except 1) needs a rebase and 2) the clang-format complaint.

kroening commented:
Both done!

martin-cs commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -626,7 +627,18 @@ int goto_analyzer_parse_optionst::perform_analysis(const optionst &options)
     if(options.get_bool_option("show"))
     {
-      static_show_domain(goto_model, *analyzer, options, out);
+      if(
```
I think I would prefer a new option for output "text-on-source" or something like that and have ```show_on_source``` just called for that.  That way we maintain backwards compatibility.

kroening replied:
Ok, it's now --show-on-source.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: cf041448d192a619f42e966967136d89c492224f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89636240
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -191,6 +192,11 @@ void goto_analyzer_parse_optionst::get_command_line_options(optionst &options)
     options.set_option("show", true);
     options.set_option("general-analysis", true);
   }
+  else if(cmdline.isset("show-on-source"))
```
Please add the option to the `--help` output.

kroening replied:
Done

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: d5d204ee68e9e96d9cc10e96c07853ad5910b9e7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90749109

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.h:
```diff
@@ -145,7 +145,7 @@ class optionst;
   "(constants)" \
   "(dependence-graph)" \
-  "(show)(verify)(simplify):" \
+  "(show)(verify)(simplify):(show-on-source)" \
```
```suggestion
  "(show)(verify)(simplify):" \
  "(show-on-source)" \
```

tautschnig commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -800,7 +811,8 @@ void goto_analyzer_parse_optionst::help()
     "\n"
     "Task options:\n"
-    " --show                       display the abstract domains\n"
+    " --show                       display the abstract states on the goto program\n"
```
```suggestion
    // NOLINTNEXTLINE(whitespace/line_length)
    " --show                       display the abstract states on the goto program\n"
```

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: 5dd37608c16ca2fd8223f94f787ed5c2a939b995).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90819223

tautschnig committed:
Update src/goto-analyzer/goto_analyzer_parse_options.h

Co-Authored-By: kroening <kroening@cs.ox.ac.uk>

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 36581670c688bd4a80f4c159c26eef01c9abc064).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90955098

chrisr-diffblue reviewed (approved):
Looks good, and a nice feature.

martin-cs commented on src/goto-analyzer/show_on_source.cpp:
```diff
@@ -0,0 +1,145 @@
+{
+  const auto abstract_state = ai.abstract_state_before(loc);
+  if(abstract_state->is_top())
+    return {};
```
This doesn't match the name of the function or the documentation; why should the file name depend on the domain?

martin-cs reviewed (approved):
I'm happy with the way this is implemented.  I'm ... undecided on the change in the usage model.  My understanding is that the usage model of CPROVER has been very much pipeline : you need the source to create the goto-program but after that you don't need it any more.  This then loops back around and changes the relationship between analysis results and code.  This is not a bad thing but it is a change of abstraction / architecture.  If we are going to go this way (and I am not for or against doing so at this stage) then we should probably think about the consequences / necessary architectural changes.  For example, do we need to store / be able to do byte-exact reconstruction of source from goto-programs?

So, I think the answer is merge this and if we find that we're adding lots more features, think about changing the architecture.

kroening commented:
Yes, we do want pipeline! However, note that "view abstract fixedpoint on source code" is still very much a feature aimed at developers of analysis tools, as opposed to users, so if this gets broken by some transformation it's not a big deal. You'll then have to view using the goto-program, which is what we currently do.
